add e2e docs

Ghostvv opened the pull request:
**Proposed changes**:
- describe how to add user text to stories
- describe how to add bot text to stories
- rename some ted default params
- describe ted's new params in the docs

**Status (please check what you already did)**:
- [ ] added some tests for the functionality
- [x] updated the documentation
- [x] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/master/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

tabergma commented on docs/docs/training-data-format.mdx:
```diff
@@ -380,26 +380,26 @@ stories:
-  - A [form](#forms)
-  - A [slot was set](#slots) event
-  - A [checkpoint](#checkpoints), which connects the story to another story
+  - A [user message](#user-messages), represented by **intent**, **entities** or **user**.
```
Shouldn't it be `intent and entities` or `user`?

tabergma commented on docs/docs/training-data-format.mdx:
````diff
@@ -458,6 +493,35 @@ stories:
 ```
 
+If you want to put bot utterance directly in the stories, use `bot:` key followed by
+the text that you want your bot to say
````
```suggestion
the text that you want your bot to say.
```

tabergma commented on docs/docs/training-data-format.mdx:
```diff
@@ -458,6 +493,35 @@ stories:
+If you have feedback (positive or negative) please share it with us on the [Rasa Forum](https://forum.rasa.com).
+
+:::
+
```
Let's add a sentence that an end-to-end story might look like this. E.g. it can be a mixture of different keys.

tabergma commented on docs/docs/training-data-format.mdx:
````diff
@@ -428,6 +428,41 @@ policies learn to predict the next action based on a *combination* of
+    - user: the actual text of the user message  # Required
+```
+
+In addition, you can add entity tags that will be extracted by end-to-end policy.
````
```suggestion
In addition, you can add entity tags that will be extracted by the end-to-end policy.
```

tabergma commented on docs/docs/training-data-format.mdx:
```diff
@@ -428,6 +428,41 @@ policies learn to predict the next action based on a *combination* of
+
+In addition, you can add entity tags that will be extracted by end-to-end policy.
+The syntax for entity tags is the same as in
+[the entities in the NLU training data](./training-data-format.mdx#entities)
```
[the NLU training data](./training-data-format.mdx#entities)

tabergma commented on docs/docs/training-data-format.mdx:
```diff
@@ -428,6 +428,41 @@ policies learn to predict the next action based on a *combination* of
+In addition, you can add entity tags that will be extracted by end-to-end policy.
+The syntax for entity tags is the same as in
+[the entities in the NLU training data](./training-data-format.mdx#entities)
+For example, to represent the sentence
+`I want to check my credit balance`, where `credit` is an entity:
```
```suggestion
For example, the following stories contains the user utterance
`I want to check my credit balance`. By using the syntax from the NLU training data `[credit](account_type)`, we mark `credit` as an entity of type `account_type`.
```

Ghostvv commented:
@tabergma do you remember if we deprecated smth that we need to put in migration guide?

Ghostvv commented:
@tabergma could you please review again. I uncovered a lot of small bugs related to updating config

tabergma commented on docs/docs/policies.mdx:
```diff
@@ -157,13 +164,23 @@ If you want to fine-tune your model, start by modifying the following parameters
-* `number_of_transformer_layers`:
-  This parameter sets the number of transformer layers to use (default: `1`).
-  The number of transformer layers corresponds to the transformer blocks to use for the model.
+* `sequence_number_of_transformer_layers`:
```
I am confused. This does not match the with the changes you mentioned earlier or am I missing something?

> number_of_transformer_layers -> dictionary “number_of_transformer_layers” with keys “text”, “action_text”, “label_action_text”, “dialogue”

tabergma replied:
same for the others

tabergma replied:
Ok, I think you simply forgot to update these here.

Ghostvv replied:
yep simply forgot to remove

tabergma commented on docs/docs/stories.mdx:
```diff
@@ -45,6 +45,19 @@ to predict the next action based on a *combination* of both the intent and
+We call it end-to-end training because actual text is used for predictions.
+
+:::caution
+This feature is experimental.
```
Are we actually raising a warning from inside the code when users are using this feature?

Ghostvv replied:
hm, I guess no. I think we should do it when we know that stories are `e2e`. @joejuzl I think you added this check somewhere, could you please add the warning

tabergma commented on docs/docs/training-data-format.mdx:
```diff
@@ -428,6 +428,42 @@ policies learn to predict the next action based on a *combination* of
+In addition, you can add entity tags that will be extracted by the end-to-end policy.
+The syntax for entity tags is the same as in
+[the NLU training data](./training-data-format.mdx#entities)
+For example, the following stories contains the user utterance
```
```suggestion
For example, the following story contains the user utterance
```

tabergma commented on rasa/utils/train_utils.py:
```diff
@@ -155,40 +161,83 @@ def _replace_deprecated_option(
+            warn_until_version=warn_until_version,
+        )
+        return {new_option: config[old_option]}
+    else:
```
We don't need this else as we return before.

tabergma commented on rasa/utils/train_utils.py:
```diff
@@ -233,3 +282,21 @@ def entity_label_to_tags(
+
+def override_defaults(
+    defaults: Optional[Dict[Text, Any]], custom: Optional[Dict[Text, Any]]
+) -> Dict[Text, Any]:
```
docstring

tabergma commented on rasa/utils/train_utils.py:
```diff
@@ -233,3 +282,21 @@ def entity_label_to_tags(
+    defaults: Optional[Dict[Text, Any]], custom: Optional[Dict[Text, Any]]
+) -> Dict[Text, Any]:
+    if defaults:
+        cfg = copy.deepcopy(defaults)
```
Please rename `cfg` to `config`.

tabergma commented on changelog/7496.improvement.md:
```diff
@@ -0,0 +1,35 @@
+Make TED an end-to-end policy. Namely, make it possible to train TED on stories that contain
```
Should we maybe add a sentence that TED can still be used in the "old" way and that nothing changed for that?

tabergma reviewed (commented):
Added a few more comment. Cannot think of anything that we deprecated right now. But will go through the code just to be sure.

tabergma commented:
Regarding deprecation: The StateFeaturizers were deprecated in the last PR already, correct?

When you look at the code changes, we have the following in a changelog entry
> `Domain.random_template_for` is deprecated and will be removed in Rasa Open Source 
3.0.0. You can alternatively use the `TemplatedNaturalLanguageGenerator`.
`Domain.action_names` is deprecated and will be removed in Rasa Open Source 
3.0.0. Please use `Domain.action_names_or_texts` instead.

So, we might want to add that to the migration guide.

Apart from that: Did we update the minimum compatible version? The changes are model breaking.

Ghostvv commented:
@wochinge could you please upgrade `migration guide` with domain changes outlined above

erohmensing commented on docs/docs/stories.mdx:
```diff
@@ -18,23 +18,23 @@ Here's an example of a dialogue in the Rasa story format:
 stories:
```
Above this under "format", we say that the inputs are expressed as intents and entities, and responses/actions as action names, which is not completely true anymore

erohmensing commented on docs/docs/stories.mdx:
```diff
@@ -45,6 +45,19 @@ to predict the next action based on a *combination* of both the intent and
+the text of the user message.
+We call it end-to-end training because actual text is used for predictions.
+
+:::caution
+This feature is experimental.
```
```suggestion
:::caution experimental feature
End-to-end training is an experimental feature.
```

erohmensing commented on docs/docs/stories.mdx:
```diff
@@ -18,23 +18,23 @@ Here's an example of a dialogue in the Rasa story format:
 ### User Messages
 
-While writing stories, you do not have to deal with the specific contents of
+While writing stories, you have two options.
```
```suggestion
While writing stories, you have two options to represent user input.
```

erohmensing commented on docs/docs/stories.mdx:
````diff
@@ -18,23 +18,23 @@ Here's an example of a dialogue in the Rasa story format:
-    - cuisine: "spanish"
+  - user: I would like [spanish](cuisine). # actual text that user input
   - action: utter_ask_num_people
 ```
 
````
Please add a link to the training data format page here for more specifics

erohmensing commented on docs/docs/stories.mdx:
```diff
@@ -54,6 +67,15 @@ You can use a response from your domain as an action by listing it as one
 in a story. Similarly, you can indicate that a story should call a custom action by including
 the name of the custom action from the `actions` list in your domain.
 
+You can also put bot utterance directly in the stories.
+
+:::caution
+This feature is experimental.
+We introduce experimental features to get feedback from our community, so we encourage you to try it out!
+However, the functionality might be changed or removed in the future.
+If you have feedback (positive or negative) please share it with us on the [Rasa Forum](https://forum.rasa.com).
+
+:::
```
I would also put this under 2 options so that users know that it's only the 2nd option that is experimental. Also rephrase the warning as i've done above on the other one

erohmensing commented on docs/docs/training-data-format.mdx:
```diff
@@ -380,26 +380,26 @@ stories:
-  - A [form](#forms)
-  - A [slot was set](#slots) event
-  - A [checkpoint](#checkpoints), which connects the story to another story
+  - A [user message](#user-messages), represented by **intent** and **entities** or **user**.
```
```suggestion
  - A [user message](#user-messages), represented by a combination of **intent** and **entities**, or **user**.
```

Some way of clarifying which two go together, doesn't have to be phrased exactly like this

erohmensing commented on docs/docs/training-data-format.mdx:
```diff
@@ -380,26 +380,26 @@ stories:
 
-All user messages are specified with the `intent:`
-key and an optional `entities:` key.
+User messages are specified with the `intent:`
+key and an optional `entities:` key or with `user:` key.
```
```suggestion
There are two ways to specify user messages. You can provide the `intent:`
key and an optional `entities:` key, or you can provide the `user:` key.
```

erohmensing commented on docs/docs/training-data-format.mdx:
```diff
@@ -428,6 +428,42 @@ policies learn to predict the next action based on a *combination* of
+If you use `user:` key while writing stories, you do not have to deal with the specific
+intents of the messages that are extracted by the NLU pipeline. Instead, you can simply put
+the text of the user message.
+We call it end-to-end training because actual text is used for predictions.
```
What I don't understand as a user is how this affects the story prediction. How is it  decided if a user message is close enough to the one ive provided to follow that story? Is it essentially "whatever this message would be classified as intent-wise based on my other NLU data (does this entry get added to training data as well?), if i get another message that falls in that intent, then this story is relevant?"

I guess this format only applies to TED so maybe that doesn't make sense, but I think we could use a bit of clarification here

erohmensing replied:
Also, is this key allowed in rules, and if not, will it error as expected?

Ghostvv replied:
you can have it rules. then it'll be string matching. I keep the idea that Rules is exactly the same as stories

Ghostvv replied:
added explanations

erohmensing commented on docs/docs/training-data-format.mdx:
````diff
@@ -458,6 +494,37 @@ stories:
   - action: action_store_feedback
 ```
 
+If you want to put bot utterance directly in the stories, use `bot:` key followed by
+the text that you want your bot to say.
````
Similarly, I have no idea how this works. Does ted just make predictions ranking "utter_greet" against "How are you?" in a greet bot and go from there? If I have some stories where I use "utter_howareyou" and others where I provide the utterance directly, is that going to confuse TED?

Ghostvv replied:
added explanations

wochinge commented on changelog/7496.improvement.md:
```diff
@@ -0,0 +1,36 @@
+Make TED an end-to-end policy. Namely, make it possible to train TED on stories that contain
```
let's also link the official docs page for it in the changelog.

Ghostvv replied:
how to add doc link in changelog? I didn't find it in the description

wochinge replied:
Should be something like `(my link test)[policies.mdx#some-section]`

Ghostvv replied:
will it know address of `policies.mdx` form where changelog is?

wochinge replied:
Yes 👍🏻  If not -> CI fails

wochinge commented on docs/docs/stories.mdx:
```diff
@@ -18,23 +18,23 @@ Here's an example of a dialogue in the Rasa story format:
-  - intent: inform
-    entities:
-    - cuisine: "spanish"
+  - user: I would like [spanish](cuisine). # actual text that user input
```
@erohmensing What do you think of adding an explanation why we need to annotate entities like this in stories?

erohmensing replied:
Yeah that would be helpful -- as you don't have to annotate the intent, i could see the difference there being confusing

Ghostvv replied:
> why we need

what do you mean by `need`, the need comes from users

wochinge replied:
we need it as we use it to train NLU with it, no?

Ghostvv replied:
we don't train NLU with it

Ghostvv replied:
I wrote in the docs `In addition, you can add entity tags that will be extracted by the end-to-end policy.` I didn't write it'll be used in NLU

Ghostvv replied:
rephrased it to "In addition, you can add entity tags that can be extracted by the `TEDPolicy`." so that it is explicit

wochinge commented on docs/docs/stories.mdx:
```diff
@@ -18,23 +18,23 @@ Here's an example of a dialogue in the Rasa story format:
+  - bot: On it                             # actual text that bot can output
   - action: utter_ask_cuisine
-  - intent: inform
-    entities:
-    - cuisine: "spanish"
+  - user: I would like [spanish](cuisine). # actual text that user input
```
considering this an experimental feature I'm wondering if it's worth adding this to the first example on this page here. In my opinion the introduction to the format should be clear and concise while covering the essential use cases. We can add formats for further usages in further sections and making it clear that this is an experimental feature.

wochinge replied:
I suggest having an additional section with all end-to-end things (similar as we have it for `checkpoints` or `OR statements`) instead of mixing it.

Ghostvv replied:
reverted this example

Ghostvv replied:
I don't think additional section will make it better, because you can mix them. I think current structure in `User Messages` and `Actions` is better

wochinge commented on docs/docs/stories.mdx:
```diff
@@ -18,23 +18,23 @@ Here's an example of a dialogue in the Rasa story format:
-While writing stories, you do not have to deal with the specific contents of
+While writing stories, you have two options to represent user input.
+
+1) You do not have to deal with the specific contents of
```
Is this actually rendering as proper MD list?

Ghostvv replied:
update so it is the same as other lists

wochinge commented on docs/docs/training-data-format.mdx:
```diff
@@ -428,6 +428,42 @@ policies learn to predict the next action based on a *combination* of
+If you use `user:` key while writing stories, you do not have to deal with the specific
+intents of the messages that are extracted by the NLU pipeline. Instead, you can simply put
+the text of the user message.
+We call it end-to-end training because actual text is used for predictions.
```
If we call it that, then we should also add it to the glossary (are we in sync with marketing / CSE about the naming?)

Ghostvv replied:
I don't think we have a name. It's rather name widely used in the field

wochinge commented on docs/docs/training-data-format.mdx:
```diff
@@ -428,6 +428,42 @@ policies learn to predict the next action based on a *combination* of
+stories:
+- story: user message structure
+  steps:
+    - user: the actual text of the user message  # Required
```
what does `# required` mean? What is it supposed to tell the user?

Ghostvv replied:
I don't know, it was there

Ghostvv replied:
removed it

wochinge commented on docs/docs/training-data-format.mdx:
```diff
@@ -428,6 +428,42 @@ policies learn to predict the next action based on a *combination* of
+[the NLU training data](./training-data-format.mdx#entities)
+For example, the following story contains the user utterance
+`I want to check my credit balance`. By using the syntax from the NLU training data
+`[credit](account_type)`, we mark `credit` as an entity of type `account_type`.
```
Not an expert on doc guidelines, but according to [this](https://www.notion.so/rasa/Tone-and-language-guidelines-cd96f21292b24c1ab7f7457c7973ec69#766266eaf5744ca094bd247409684d28) we should say `you` (#soMeta)

Ghostvv replied:
where?

wochinge commented on rasa/utils/train_utils.py:
```diff
@@ -155,40 +161,87 @@ def _replace_deprecated_option(
+            f"Option '{old_option}' got renamed to '{new_option}'. "
+            f"Please update your configuration file.",
+            warn_until_version=warn_until_version,
+        )
```
why are their code changes in a documentation PR?

Ghostvv replied:
because documentation needs config changes, config changes needs code changes. I cannot switch back and forth

Ghostvv commented:
@wochinge @erohmensing I think `e2e` requires separate documentation page. But currently we don't know enough about it to create a page where we could recommend smth. I suggest postpone it post release

erohmensing commented:
What don't you know about, e2e? 😅 If you want it to have its own page, don't worry yet about where the content should go, we can figure that out later, but I wouldn't say it should wait til post-release. If its important and/or confusing enough that it needs its own page to explain to users, they should have that information available when the feature is available

Ghostvv commented:
> What don't we know enough about, e2e?

this is the thing, nothing except that you can put text in the training data, which I already added. I also added more according to your comments. Is it good to go then?

erohmensing commented:
I'm confused, you implemented the feature, surely you know all about it? In any case after talking with @akelad I think she made a great point -- we don't want to recommend this in production, if its super experimental it makes more sense for it to live on its own page, and only mentioning it on the "writing stories" and "training data format" pages as a note that leads you to that feature page. What do you think?

Ghostvv commented:
> surely you know all about it?

what do you mean by all?
- How it works? but it is literally training data format changes and slightly more complicated architecture, but we don't describe NN architectures in docs, at least we didn't do it before.

- How to best use it? I have no idea. My usage experience is limited to `e2ebot` 😅

I described all I know about it in the changes, except detailed architecture

> only mentioning it on the "writing stories" and "training data format" pages as a note

I'm confused. Don't mention it in `training-data-format.mdx`?

erohmensing commented:
Sorry, my bad, I assumed that in the training data format. Okay, so we just don't know what to tell users about it 🙈 i'm not sure then, haha

Ghostvv commented:
yes we can tell them how to use it (in terms of how to create training data and hit train button) but we cannot really tell How (best) to use it 😂

amn41 commented:
had a quick look at these docs - I think we should stick with how we previously described stories, and then add an additional section at the end of the page about end-to-end.

Ghostvv commented:
so in both files? In description of training data format as well?

amn41 commented:
yes I think so

akelad commented:
i talked with ella about this on friday as well, and I agree with Alan that we should separate out e2e from the regular content. But do we really need descriptions of it on both the training data format and the stories page?

Ghostvv commented:
> But do we really need descriptions of it on both the training data format and the stories page?

Just wanted to ask that. But in general these two pages overlap a lot. I think it makes sense to do the same for e2e, maybe in training data format it'll be about syntax, and in stories I'll answer the questions that Ella raised that NLU is ignored

akelad commented:
ok i think actually some of the content from training data format needs to be pulled out from there and moved to stories (or it's already moved to stories). We must have missed removing some of it during the docs rework. I would say maybe just add a shorter section about e2e in training data format, and explain the full thing in stories? Or should it actually have it's own page? Since it's a story and also nlu data

Ghostvv commented on docs/docs/training-data-format.mdx:
```diff
@@ -380,26 +380,26 @@ stories:
 
 Each step can be one of the following:
 
-  - A [user message](#user-messages), represented by **intent** and **entities**.
-  - An [or statement](#or-statement), which includes two or more user messages under it
-  - A bot [action](#actions)
-  - A [form](#forms)
-  - A [slot was set](#slots) event
-  - A [checkpoint](#checkpoints), which connects the story to another story
```
@akelad @amn41 for my understanding: do you propose to return that to original state as well? So we present data format as if e2e doesn't exist?

akelad replied:
yeah i don't think we should have the end to end format on an overview level if we still sell it as an experimental feature

Ghostvv replied:
ok, thank you for the clarification. Reworking my changes

amn41 replied:
yes I think that's what we should do

Ghostvv commented:
@erohmensing @akelad @amn41 @b-quachtran created separate paragraph

b-quachtran commented on docs/docs/migration-guide.mdx:
```diff
@@ -12,8 +12,25 @@ how you can migrate from one version to another.
 
+### General
+
+Some model options for `TEDPolicy` got renamed.
```
```suggestion
`TEDPolicy`'s  `transformer_size`, `number_of_transformer_layers`, and `dense_dimensions` parameters have been renamed.
```

b-quachtran commented on docs/docs/migration-guide.mdx:
```diff
@@ -12,8 +12,25 @@ how you can migrate from one version to another.
+Some model options for `TEDPolicy` got renamed.
+Please update your configuration files using the following mapping:
+
+|      Old model option       |                  New model option                      |
```
```suggestion
|      Old Model Parameter    |                 New Model Parameter                    |
```

b-quachtran commented on docs/docs/migration-guide.mdx:
```diff
@@ -12,8 +12,25 @@ how you can migrate from one version to another.
+
+|      Old model option       |                  New model option                      |
+|-----------------------------|--------------------------------------------------------|
+|transformer_size             |dictionary “transformer_size” with keys                 |
```
Can you add an example dict for each of the new model parameters?

akelad commented on docs/docs/stories.mdx:
````diff
@@ -32,6 +33,9 @@ stories:
 ```
 
+See the [training data format](./training-data-format.mdx#stories)
+for detailed description of how to write stories.
````
🤔 why?

Ghostvv replied:
I was asked in one of the comments above. I can remove that

akelad commented on docs/docs/stories.mdx:
```diff
@@ -185,3 +190,64 @@ into a single file for evaluation. Read more about this format in [Testing Your
+End-to-end training is an experimental feature.
+We introduce experimental features to get feedback from our community, so we encourage you to try it out!
+However, the functionality might be changed or removed in the future.
+If you have feedback (positive or negative) please share it with us on the [Rasa Forum](https://forum.rasa.com).
```
Might make sense to actually create a specific forum thread for it

akelad commented on docs/docs/training-data-format.mdx:
````diff
@@ -458,6 +458,7 @@ stories:
   - action: action_store_feedback
 ```
 
+
````
pls delete

akelad commented on docs/docs/training-data-format.mdx:
```diff
@@ -685,3 +686,72 @@ rasa test
 [Testing Your Assistant](testing-your-assistant.mdx).
+
+
+## End-to-end Training Stories
```
```suggestion
## End-to-end Training
```
since we call it that on the other page, and it also fits in the sidebar better

akelad commented on docs/docs/stories.mdx:
```diff
@@ -185,3 +190,64 @@ into a single file for evaluation. Read more about this format in [Testing Your
+:::note increased train time
+End-to-end training requires significantly more parameters in `TEDPolicy`.
+Therefore, training an end-to-end model might require significant computational
+resources depending on how many end-to-end turns you have in your stories.
```
there's too many notes on this page imo, if it's important enough, it should just go in the text of the docs

Ghostvv replied:
removed notes

akelad commented on docs/docs/stories.mdx:
```diff
@@ -45,6 +49,7 @@ to predict the next action based on a *combination* of both the intent and
 entities (you can, however, change this behavior using the
 [use_entities](./domain.mdx#ignoring-entities-for-certain-intents) attribute).
 
+
```
pls remove

akelad reviewed (commented):
i only had time to skim these docs, I think generally they probably need a lot more work, but i don't have the capacity to review them in detail atm

akelad commented on docs/docs/training-data-format.mdx:
```diff
@@ -685,3 +686,72 @@ rasa test
+stories:
+- story: user message structure
+  steps:
+    - user: the actual text of the user message
```
should there be an action after this ?

Ghostvv replied:
yes but in the example for intents on lines 404-410 there is no actions

Ghostvv replied:
I just copy pasted

wochinge replied:
Ok, then how about adding an action here?

Ghostvv replied:
done

wochinge commented on docs/docs/stories.mdx:
```diff
@@ -185,3 +187,64 @@ into a single file for evaluation. Read more about this format in [Testing Your
+
+With end-to-end training, you do not have to deal with the specific
+intents of the messages that are extracted by the NLU pipeline
+or with separate `utter_` templates in the domain file.
```
they are now called `responses`
```suggestion
or with separate `utter_` responses in the domain file.
```

wochinge commented on docs/docs/stories.mdx:
```diff
@@ -185,3 +187,64 @@ into a single file for evaluation. Read more about this format in [Testing Your
+With end-to-end training, you do not have to deal with the specific
+intents of the messages that are extracted by the NLU pipeline
+or with separate `utter_` templates in the domain file.
+Instead, you can simply put the texts of the user and/or bot messages directly in the stories.
```
```suggestion
Instead, you can put the texts of the user and/or bot messages directly in the stories.
```

wochinge commented on docs/docs/stories.mdx:
```diff
@@ -185,3 +187,64 @@ into a single file for evaluation. Read more about this format in [Testing Your
+See the [training data format](./training-data-format.mdx#end-to-end-training-stories)
+for detailed description of how to write end-to-end stories.
+
+Rasa end-to-end training is fully integrated with standard Rasa approach.
+It means that you can have mixed stories with some steps defined by actions
+and other steps defined by user or bot utterance directly.
```
?
```suggestion
You can mix training data in end-to-end format with labeled training data which has `intent`s and `action`s specified. This means stories can have steps defined by actions
and other steps defined by user or bot utterances directly.
```

wochinge commented on docs/docs/stories.mdx:
```diff
@@ -185,3 +187,64 @@ into a single file for evaluation. Read more about this format in [Testing Your
+
+:::note nlu is ignored during e2e
+For end-to-end user inputs, intents classified by the NLU pipeline
+and extracted entities are ignored.
```
I'd add that they should still provide NLU training data. Users might otherwise think they can ignore NLU completely.

Ghostvv replied:
what do you mean? if they don't have intents in the stories, they don't need NLU data

wochinge replied:
Mhm, but isn't the `CountVectorizer`'s vocabulary relatively small (unless they have very very many end-to-end stories)? Having NLU training data seems the easier way to provide the data to me?

Ghostvv replied:
not sure I understand what do you mean? the fact that you'll increase `CVF`'s vocabulary with NLU data will not affect TEDPolicy training

Ghostvv replied:
NLU and e2e core is effectively independent, the only reason we reuse featurization is to avoid duplication, so data in NLU doesn’t influence e2e core model and data in e2e core doesn’t influence NLU model.

wochinge commented on docs/docs/stories.mdx:
```diff
@@ -185,3 +187,64 @@ into a single file for evaluation. Read more about this format in [Testing Your
+  rules based on user text will only match if the user
+  text strings inside your rules and input during prediction are identical.
+
+- `TEDPolicy` passes user text through additional Neural Network to create
```
```suggestion
- `TEDPolicy` passes user text through an additional Neural Network to create
```

wochinge commented on docs/docs/stories.mdx:
```diff
@@ -185,3 +187,64 @@ into a single file for evaluation. Read more about this format in [Testing Your
+  text strings inside your rules and input during prediction are identical.
+
+- `TEDPolicy` passes user text through additional Neural Network to create
+  hidden representation of the text. In order to obtain robust performance you
```
```suggestion
  hidden representations of the text. In order to obtain robust performance you
```

wochinge commented on docs/docs/stories.mdx:
```diff
@@ -185,3 +187,64 @@ into a single file for evaluation. Read more about this format in [Testing Your
+  need to create sufficient number of stories that have different user texts
+  during this dialogue turn.
+
+:::note next utterance selection
```
I think this should be a section instead of a note.

wochinge commented on docs/docs/stories.mdx:
```diff
@@ -185,3 +187,64 @@ into a single file for evaluation. Read more about this format in [Testing Your
+In case of a template, `TEDPolicy` sees only the name of the template, while
+if you provide actual utterance using `bot:` key,
+`TEDPolicy` will featurize it as textual input depending on the NLU configuration.
+Which can help in case of similar utterances in slightly different situations.
```
```suggestion
This can help in case of similar utterances in slightly different situations.
```

wochinge commented on docs/docs/stories.mdx:
```diff
@@ -185,3 +187,64 @@ into a single file for evaluation. Read more about this format in [Testing Your
+if you provide actual utterance using `bot:` key,
+`TEDPolicy` will featurize it as textual input depending on the NLU configuration.
+Which can help in case of similar utterances in slightly different situations.
+But can also make things harder to learn.
```
```suggestion
However, this can also make things harder to learn because ....
```

wochinge commented on docs/docs/stories.mdx:
```diff
@@ -185,3 +187,64 @@ into a single file for evaluation. Read more about this format in [Testing Your
+
+:::note next utterance selection
+Rasa policies are trained for next utterance selection.
+The only difference to creating `utter_` template is how `TEDPolicy` featurizes
```
```suggestion
The only difference to creating `utter_` response is how `TEDPolicy` featurizes
```

wochinge commented on docs/docs/stories.mdx:
```diff
@@ -185,3 +187,64 @@ into a single file for evaluation. Read more about this format in [Testing Your
+Rasa policies are trained for next utterance selection.
+The only difference to creating `utter_` template is how `TEDPolicy` featurizes
+bot utterances.
+In case of a template, `TEDPolicy` sees only the name of the template, while
```
```suggestion
In case of a template, `TEDPolicy` sees only the name of the template, while
```
```suggestion
In case of an `utter_` action, `TEDPolicy` sees only the name of the action, while
```

wochinge commented on docs/docs/training-data-format.mdx:
```diff
@@ -381,11 +381,11 @@ stories:
-  - A [slot was set](#slots) event
-  - A [checkpoint](#checkpoints), which connects the story to another story
+  - An [or statement](#or-statement), which includes two or more user messages under it.
+  - A bot [action](#actions).
```
```suggestion
  - A bot [action](#actions).
  - A bot [utterance](#end-to-end-training)
```

Ghostvv replied:
this is left the same intentionally, see discussion above

wochinge commented on docs/docs/training-data-format.mdx:
```diff
@@ -381,11 +381,11 @@ stories:
 Each step can be one of the following:
 
   - A [user message](#user-messages), represented by **intent** and **entities**.
```
```suggestion
  - A [user message](#user-messages), represented by **intent** and **entities** or by a text.
```

Ghostvv replied:
this is left the same intentionally, see discussion above

wochinge commented on docs/docs/training-data-format.mdx:
```diff
@@ -685,3 +685,72 @@ rasa test
+
+
+Similarly you can put bot utterances directly in the stories,
+use `bot:` key followed by the text that you want your bot to say.
```
```suggestion
by using the `bot:` key followed by the text that you want your bot to say.
```

wochinge commented on docs/docs/training-data-format.mdx:
````diff
@@ -685,3 +685,72 @@ rasa test
+  - bot: Hello
+```
+
+You can also have full end-to-end story:
````
How about adding a story which mixes all 4 things (`user:`, `intent`, `action`, `bot`)?

wochinge commented on docs/docs/training-data-format.mdx:
```diff
@@ -685,3 +685,72 @@ rasa test
+
+:::
+
+With end-to-end training, you do not have to deal with the specific
```
```suggestion
With [end-to-end training](stories.mdx#end-to-end-training), you do not have to deal with the specific
```

wochinge reviewed (commented):
Great second iteration! Left some comments - I think somebody from CSE should give the final 👍🏻

samsucik commented on docs/docs/migration-guide.mdx:
```diff
@@ -12,8 +12,26 @@ how you can migrate from one version to another.
+
+|      Old Model Parameter    |                 New Model Parameter                    |
+|-----------------------------|--------------------------------------------------------|
+|transformer_size             |dictionary “transformer_size” with keys                 |
+|                             |“text”, “action_text”, “label_action_text”, “dialogue”  |
+|number_of_transformer_layers |dictionary “number_of_transformer_layers” with keys     |
+|                             |“text”, “action_text”, “label_action_text”, “dialogue”  |
+|dense_dimension              |dictionary “dense_dimension” with keys                  |
+|                             |“text”, “action_text”, “label_action_text”, “intent”,   |
+|                             |“action_name”, “label_action_name”, “entities”, “slots”,|
+|                             |“active_loop”                                           |
```
```suggestion
|`transformer_size`             |dictionary `transformer_size` with keys                 |
|                             |`text`, `action_text`, `label_action_text`, `dialogue`  |
|`number_of_transformer_layers` |dictionary `number_of_transformer_layers` with keys     |
|                             |`text`, `action_text`, `label_action_text`, `dialogue`  |
|`dense_dimension`              |dictionary `dense_dimension` with keys                  |
|                             |`text`, `action_text`, `label_action_text`, `intent`,   |
|                             |`action_name`, `label_action_name`, entities`, `slots`,|
|                             |`active_loop`                                           |
```

samsucik commented on docs/docs/migration-guide.mdx:
```diff
@@ -24,6 +42,12 @@ Support for Markdown data will be removed entirely in Rasa Open Source 3.0.0.
 Please convert your existing Markdown data by using the commands
 described [here](./migration-guide.mdx#training-data-files).
 
+#### Others
```
```suggestion
#### Other
```

samsucik commented on docs/docs/policies.mdx:
```diff
@@ -87,18 +87,25 @@ The Transformer Embedding Dialogue (TED) Policy is described in
 
-1. Concatenate user input (user intent and entities), previous system actions, slots and active forms for each time
-  step into an input vector to pre-transformer embedding layer.
+1. Concatenate user input (user intent and entities) or user text processed through a user sequence transformer encoder,
+   previous system actions or bot utterances processed through a bot sequence transformer encoder,
+   slots and active forms for each time
+   step into an input vector to pre dialogue transformer embedding layer.
```
```suggestion
1. Concatenate features for
   - user input (user intent and entities) or user text processed through a user sequence transformer encoder,
   - previous system actions or bot utterances processed through a bot sequence transformer encoder,
   - slots and active forms

   for each time step into an input vector to the embedding layer that precedes the 
   dialogue transformer.
```

samsucik commented on docs/docs/policies.mdx:
```diff
@@ -87,18 +87,25 @@ The Transformer Embedding Dialogue (TED) Policy is described in
+   step into an input vector to pre dialogue transformer embedding layer.
 
-2. Feed the input vector into a transformer.
+2. Feed the input vector into a dialogue transformer encoder.
```
```suggestion
2. Feed the embedding of the input vector into the dialogue transformer encoder.
```

samsucik commented on docs/docs/policies.mdx:
```diff
@@ -87,18 +87,25 @@ The Transformer Embedding Dialogue (TED) Policy is described in
+2. Feed the input vector into a dialogue transformer encoder.
 
-3. Apply a dense layer to the output of the transformer to get embeddings of a dialogue for each time step.
+3. Apply a dense layer to the output of the dialogue transformer to get embeddings of a dialogue for each time step.
```
```suggestion
3. Apply a dense layer to the output of the dialogue transformer to get embeddings of the dialogue for each time step.
```

Ghostvv commented on changelog/7496.improvement.md:
```diff
@@ -0,0 +1,36 @@
+Make [TED Policy](./policies.mdx#ted-policy) an end-to-end policy. Namely, make it possible to train TED on stories that contain
```
do we have that? ok

Ghostvv replied:
is it `7496.feature.md`

samsucik commented on docs/docs/policies.mdx:
```diff
@@ -87,18 +87,25 @@ The Transformer Embedding Dialogue (TED) Policy is described in
 5. Calculate the similarity between the dialogue embedding and embedded system actions.
    This step is based on the [StarSpace](https://arxiv.org/abs/1709.03856) idea.
 
+6. Concatenate the token output of the user sequence transformer encoder
```
```suggestion
6. Concatenate the token-level output of the user sequence transformer encoder
```

samsucik commented on docs/docs/policies.mdx:
```diff
@@ -87,18 +87,25 @@ The Transformer Embedding Dialogue (TED) Policy is described in
 This policy has a pre-defined architecture, which comprises the
 following steps:
```
I think I'm missing some high-level overview of the different things that TED does. Could we add such thing before enumerating these steps? From the detailed steps below I get the idea that TED 1) predicts next action based on the starspace-inspired similarity (though this isn't said explicitly), 2) predicts entities (in the most recent user turn?), but it requires a lot of effort to extract this high-level info from the steps.

Ghostvv replied:
copied description form DIET

samsucik commented on docs/docs/policies.mdx:
```diff
@@ -178,109 +170,143 @@ However, additional parameters exist that can be adapted.
+|                                       | label_action_text: []  | and labels. The number of hidden layers is                   |
+|                                       |                        | equal to the length of the corresponding list.               |
++---------------------------------------+------------------------+--------------------------------------------------------------+
+| dense_dimension                       | text: 128              | Dense dimension for sparse features to use.                  |
+|                                       | action_text: 128       |                                                              |
```
```suggestion
| dense_dimension                       | text: 128              | Dense dimension for sparse features to use after they are    |
|                                       | action_text: 128       | converted into dense features.                               |
```

samsucik commented on docs/docs/policies.mdx:
```diff
@@ -178,109 +170,143 @@ However, additional parameters exist that can be adapted.
+|                                       | slots: 20              |                                                              |
+|                                       | active_loop: 20        |                                                              |
++---------------------------------------+------------------------+--------------------------------------------------------------+
+| concat_dimension                      | text: 128              | Concat dimension for sequence and sentence features.         |
+|                                       | action_text: 128       |                                                              |
```
```suggestion
| concat_dimension                      | text: 128              | Common dimension to which sequence and sentence features of  |
|                                       | action_text: 128       | different dimensions get converted before concatenation.     |
```

samsucik commented on docs/docs/policies.mdx:
```diff
@@ -178,109 +170,143 @@ However, additional parameters exist that can be adapted.
+|                                       | label_action_text: 1   | Number of layers in bot text sequence transformer encoder.   |
+|                                       | dialogue: 1            | Number of layers in dialogue transformer encoder.            |
++---------------------------------------+------------------------+--------------------------------------------------------------+
+| number_of_attention_heads             | 4                      | Number of attention heads in transformers.                   |
```
```suggestion
| number_of_attention_heads             | 4                      | Number of self-attention heads in transformers.                   |
```

samsucik commented on docs/docs/policies.mdx:
```diff
@@ -178,109 +170,143 @@ However, additional parameters exist that can be adapted.
++---------------------------------------+------------------------+--------------------------------------------------------------+
+| random_seed                           | None                   | Set random seed to any 'int' to get reproducible results.    |
++---------------------------------------+------------------------+--------------------------------------------------------------+
+| embedding_dimension                   | 20                     | Dimension size of embedding vectors.                         |
```
```suggestion
| embedding_dimension                   | 20                     | Dimension size of dialogue & system action embedding vectors.|
```

samsucik replied:
Meant to better disambiguate `embedding_dimension` and `encoding_dimension`.

samsucik commented on docs/docs/policies.mdx:
```diff
@@ -178,109 +170,143 @@ However, additional parameters exist that can be adapted.
++---------------------------------------+------------------------+--------------------------------------------------------------+
+| use_sparse_input_dropout              | True                   | If 'True' apply dropout to sparse input tensors.             |
++---------------------------------------+------------------------+--------------------------------------------------------------+
+| use_dense_input_dropout               | True                   | If 'True' apply dropout to dense input tensors.              |
```
I think this isn't accurate: it describes dropout applied to _sparse_ input tensors after they are converted into dense ones...

Ghostvv replied:
good catch

samsucik commented on docs/docs/policies.mdx:
```diff
@@ -178,109 +170,143 @@ However, additional parameters exist that can be adapted.
+|                                       |                        | Requires `evaluate_on_number_of_examples > 0` and            |
+|                                       |                        | `evaluate_every_number_of_epochs > 0`                        |
++---------------------------------------+------------------------+--------------------------------------------------------------+
+| e2e_confidence_threshold              | 0.5                    | The threshold that ensures that end-to-end is picked only if |
```
What is a reasonable value range for this one?

Ghostvv replied:
I have no idea. The logic behind `0.5` is that in this case we're sure that it is winning label

samsucik replied:
Okay then. I reckon that with e2e there will be things where we simply don't have good advice yet 🙂

Ghostvv replied:
we don't even have the bad one 😅

samsucik commented on docs/docs/stories.mdx:
```diff
@@ -185,3 +187,64 @@ into a single file for evaluation. Read more about this format in [Testing Your
+With end-to-end training, you do not have to deal with the specific
+intents of the messages that are extracted by the NLU pipeline
+or with separate `utter_` responses in the domain file.
+Instead, you can put the texts of the user and/or bot messages directly in the stories.
```
```suggestion
Instead, you can include the text of the user messages and/or bot responses directly in your stories.
```

samsucik commented on docs/docs/stories.mdx:
```diff
@@ -185,3 +187,64 @@ into a single file for evaluation. Read more about this format in [Testing Your
+intents of the messages that are extracted by the NLU pipeline
+or with separate `utter_` responses in the domain file.
+Instead, you can put the texts of the user and/or bot messages directly in the stories.
+See the [training data format](./training-data-format.mdx#end-to-end-training-stories)
```
```suggestion
See the [training data format](./training-data-format.mdx#end-to-end-training)
```

samsucik commented on docs/docs/stories.mdx:
```diff
@@ -185,3 +187,64 @@ into a single file for evaluation. Read more about this format in [Testing Your
+for detailed description of how to write end-to-end stories.
+
+Rasa end-to-end training is fully integrated with standard Rasa approach.
+It means that you can have mixed stories with some steps defined by actions
```
```suggestion
It means that you can have mixed stories with some steps defined by intents or actions
```

samsucik commented on docs/docs/stories.mdx:
```diff
@@ -185,3 +187,64 @@ into a single file for evaluation. Read more about this format in [Testing Your
+
+Rasa end-to-end training is fully integrated with standard Rasa approach.
+It means that you can have mixed stories with some steps defined by actions
+and other steps defined by user or bot utterance directly.
```
```suggestion
and other steps defined directly by user or bot utterances.
```

samsucik commented on docs/docs/stories.mdx:
```diff
@@ -185,3 +187,63 @@ into a single file for evaluation. Read more about this format in [Testing Your
+See the [training data format](./training-data-format.mdx#end-to-end-training-stories)
+for detailed description of how to write end-to-end stories.
+
+You can mix training data in end-to-end format with labeled training data which has `intent`s and `action`s specified. This means stories can have steps defined by actions
+and other steps defined by user or bot utterances directly.
```
```suggestion
You can mix training data in the end-to-end format with labeled training data which has 
`intent`s and `action`s specified: Stories can have some steps defined by intents/actions
and other steps defined directly by user or bot utterances.
```

samsucik commented on docs/docs/stories.mdx:
```diff
@@ -185,3 +187,63 @@ into a single file for evaluation. Read more about this format in [Testing Your
+You can mix training data in end-to-end format with labeled training data which has `intent`s and `action`s specified. This means stories can have steps defined by actions
+and other steps defined by user or bot utterances directly.
+
+We call it end-to-end training because actual text is fed into policies.
```
```suggestion
We call it end-to-end training because actual text can be consumed and learnt to be predicted by policies.
```

samsucik commented on docs/docs/stories.mdx:
```diff
@@ -185,3 +187,63 @@ into a single file for evaluation. Read more about this format in [Testing Your
+
+- `TEDPolicy` passes user text through an additional Neural Network to create
+  hidden representations of the text. In order to obtain robust performance you
+  need to create sufficient number of stories that have different user texts
+  during this dialogue turn.
```
```suggestion
  need to provide enough training stories to capture a variety of user texts for any given dialogue turn.
```

samsucik replied:
```suggestion
  need to provide enough training stories to capture a variety of user texts for any 
  end-to-end dialogue turn.
```

samsucik commented on docs/docs/training-data-format.mdx:
```diff
@@ -685,3 +685,72 @@ rasa test
+In addition, you can add entity tags that can be extracted
+by the [TED Policy](./policies.mdx#ted-policy).
+The syntax for entity tags is the same as in
+[the NLU training data](./training-data-format.mdx#entities)
```
```suggestion
[the NLU training data](./training-data-format.mdx#entities).
```

samsucik commented on docs/docs/training-data-format.mdx:
````diff
@@ -685,3 +685,72 @@ rasa test
+
+These end-to-end user messages follow the format:
+
+```yaml-rasa
````
```suggestion
```yaml-rasa {4}
```

samsucik commented on docs/docs/training-data-format.mdx:
````diff
@@ -685,3 +685,72 @@ rasa test
+```
+
+
+Similarly you can put bot utterances directly in the stories,
````
```suggestion
Similarly, you can put bot utterances directly in the stories,
```

samsucik commented on docs/docs/training-data-format.mdx:
```diff
@@ -685,3 +685,72 @@ rasa test
+With [end-to-end training](stories.mdx#end-to-end-training), you do not have to deal with the specific
+intents of the messages that are extracted by the NLU pipeline.
+Instead, you can simply put the text of the user message directly in the stories,
+by using `user:` key.
```
For the sake of consistency, we might want to use "key" only for the actual word, not for the following colon.

```suggestion
by using `user` key.
```

samsucik commented on docs/docs/training-data-format.mdx:
```diff
@@ -685,3 +685,72 @@ rasa test
+
+
+Similarly you can put bot utterances directly in the stories,
+by using the `bot:` key followed by the text that you want your bot to say.
```
See my comment above for `user:`.

```suggestion
by using the `bot` key followed by the text that you want your bot to say.
```

samsucik commented on docs/docs/training-data-format.mdx:
```diff
@@ -685,3 +685,72 @@ rasa test
+Similarly you can put bot utterances directly in the stories,
+by using the `bot:` key followed by the text that you want your bot to say.
+
+A story with only bot utterance might look like this:
```
```suggestion
A story with only a bot utterance might look like this:
```

samsucik commented on docs/docs/training-data-format.mdx:
````diff
@@ -685,3 +685,72 @@ rasa test
+```
+
+Rasa end-to-end training is fully integrated with standard Rasa approach.
+It means that you can have mixed stories with some steps defined by actions
+and other steps defined by user or bot utterance directly.
````
```suggestion
It means that you can have mixed stories with some steps defined by actions or intents
and other steps defined directly by user messages or bot responses.
```

samsucik commented on docs/docs/migration-guide.mdx:
```diff
@@ -12,8 +12,26 @@ how you can migrate from one version to another.
+
+|      Old Model Parameter    |                 New Model Parameter                    |
+|-----------------------------|--------------------------------------------------------|
+|`transformer_size`             |dictionary `transformer_size` with keys                 |
+|                             |`text`, `action_text`, `label_action_text`, `dialogue`  |
+|`number_of_transformer_layers` |dictionary `number_of_transformer_layers` with keys     |
+|                             |`text`, `action_text`, `label_action_text`, `dialogue`  |
+|`dense_dimension`              |dictionary `dense_dimension` with keys                  |
+|                             |`text`, `action_text`, `label_action_text`, `intent`,   |
+|                             |`action_name`, `label_action_name`, entities`, `slots`,|
+|                             |`active_loop`                                           |
```
This fixes my earlier suggestion that destroyed the table 🙂 
```suggestion
|`transformer_size`           |dictionary `transformer_size` with keys                 |
|                             |`text`, `action_text`, `label_action_text`, `dialogue`  |
|`number_of_transformer_layers`|dictionary `number_of_transformer_layers` with keys    |
|                             |`text`, `action_text`, `label_action_text`, `dialogue`  |
|`dense_dimension`            |dictionary `dense_dimension` with keys                  |
|                             |`text`, `action_text`, `label_action_text`, `intent`,   |
|                             |`action_name`, `label_action_name`, `entities`, `slots`,|
|                             |`active_loop`                                           |
```

wochinge commented on docs/docs/stories.mdx:
```diff
@@ -185,3 +187,56 @@ into a single file for evaluation. Read more about this format in [Testing Your
+and extracted entities are ignored.
+
+
+Only [Rule Policy](./policies.mdx#rule-policy)
```
should we mention that they should set `intent_tokenization_flag: True`?

Ghostvv replied:
why? they don't have to use it. It is optional

wochinge replied:
Ok - maybe my real question is: Why did we set this in the example bot?

Ghostvv replied:
tokenizing names is an option that is nice to have for advanced users if they create consistent action names, it is not required to train e2e model

it is an optional configuration decision that might improve bot performance

wochinge commented on docs/docs/training-data-format.mdx:
```diff
@@ -685,3 +685,83 @@ rasa test
+
+With [end-to-end training](stories.mdx#end-to-end-training), you do not have to deal with the specific
+intents of the messages that are extracted by the NLU pipeline.
+Instead, you can simply put the text of the user message directly in the stories,
```
```suggestion
Instead, you can put the text of the user message directly in the stories,
```

wochinge commented on docs/docs/training-data-format.mdx:
````diff
@@ -685,3 +685,83 @@ rasa test
+  - bot: Hello, Ivan!
+```
+
+You can also have full end-to-end story:
````
```suggestion
You can also have a full end-to-end story:
```

wochinge reviewed (commented):
Looks great 🪂

joejuzl commented on docs/docs/training-data-format.mdx:
```diff
@@ -685,3 +685,83 @@ rasa test
+  - intent: greet
+    entities:
+      name: Ivan
+  - bot: Hello, Ivan!
```
I wonder if this is a good example? Isn't this a situation were you would always use an `utter_` action so you could output a slot value?

Ghostvv replied:
true

Ghostvv replied:
`Hello, a person with a name!` does it sound good 😅
